<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.12">
<title>Mapping traits in biparental populations with genotyping-by-sequencing data</title>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700">
<style>
/* Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
/* Uncomment @import statement to use as custom stylesheet */
/*@import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700";*/
article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section{display:block}
audio,video{display:inline-block}
audio:not([controls]){display:none;height:0}
html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%}
a{background:none}
a:focus{outline:thin dotted}
a:active,a:hover{outline:0}
h1{font-size:2em;margin:.67em 0}
abbr[title]{border-bottom:1px dotted}
b,strong{font-weight:bold}
dfn{font-style:italic}
hr{-moz-box-sizing:content-box;box-sizing:content-box;height:0}
mark{background:#ff0;color:#000}
code,kbd,pre,samp{font-family:monospace;font-size:1em}
pre{white-space:pre-wrap}
q{quotes:"\201C" "\201D" "\2018" "\2019"}
small{font-size:80%}
sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}
sup{top:-.5em}
sub{bottom:-.25em}
img{border:0}
svg:not(:root){overflow:hidden}
figure{margin:0}
fieldset{border:1px solid silver;margin:0 2px;padding:.35em .625em .75em}
legend{border:0;padding:0}
button,input,select,textarea{font-family:inherit;font-size:100%;margin:0}
button,input{line-height:normal}
button,select{text-transform:none}
button,html input[type="button"],input[type="reset"],input[type="submit"]{-webkit-appearance:button;cursor:pointer}
button[disabled],html input[disabled]{cursor:default}
input[type="checkbox"],input[type="radio"]{box-sizing:border-box;padding:0}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}
textarea{overflow:auto;vertical-align:top}
table{border-collapse:collapse;border-spacing:0}
*,*::before,*::after{-moz-box-sizing:border-box;-webkit-box-sizing:border-box;box-sizing:border-box}
html,body{font-size:100%}
body{background:#fff;color:rgba(0,0,0,.8);padding:0;margin:0;font-family:"Noto Serif","DejaVu Serif",serif;font-weight:400;font-style:normal;line-height:1;position:relative;cursor:auto;tab-size:4;word-wrap:anywhere;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased}
a:hover{cursor:pointer}
img,object,embed{max-width:100%;height:auto}
object,embed{height:100%}
img{-ms-interpolation-mode:bicubic}
.left{float:left!important}
.right{float:right!important}
.text-left{text-align:left!important}
.text-right{text-align:right!important}
.text-center{text-align:center!important}
.text-justify{text-align:justify!important}
.hide{display:none}
img,object,svg{display:inline-block;vertical-align:middle}
textarea{height:auto;min-height:50px}
select{width:100%}
.subheader,.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{line-height:1.45;color:#7a2518;font-weight:400;margin-top:0;margin-bottom:.25em}
div,dl,dt,dd,ul,ol,li,h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6,pre,form,p,blockquote,th,td{margin:0;padding:0}
a{color:#2156a5;text-decoration:underline;line-height:inherit}
a:hover,a:focus{color:#1d4b8f}
a img{border:0}
p{font-family:inherit;font-weight:400;font-size:1em;line-height:1.6;margin-bottom:1.25em;text-rendering:optimizeLegibility}
p aside{font-size:.875em;line-height:1.35;font-style:italic}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{font-family:"Open Sans","DejaVu Sans",sans-serif;font-weight:300;font-style:normal;color:#ba3925;text-rendering:optimizeLegibility;margin-top:1em;margin-bottom:.5em;line-height:1.0125em}
h1 small,h2 small,h3 small,#toctitle small,.sidebarblock>.content>.title small,h4 small,h5 small,h6 small{font-size:60%;color:#e99b8f;line-height:0}
h1{font-size:2.125em}
h2{font-size:1.6875em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.375em}
h4,h5{font-size:1.125em}
h6{font-size:1em}
hr{border:solid #dddddf;border-width:1px 0 0;clear:both;margin:1.25em 0 1.1875em;height:0}
em,i{font-style:italic;line-height:inherit}
strong,b{font-weight:bold;line-height:inherit}
small{font-size:60%;line-height:inherit}
code{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;font-weight:400;color:rgba(0,0,0,.9)}
ul,ol,dl{font-size:1em;line-height:1.6;margin-bottom:1.25em;list-style-position:outside;font-family:inherit}
ul,ol{margin-left:1.5em}
ul li ul,ul li ol{margin-left:1.25em;margin-bottom:0;font-size:1em}
ul.square li ul,ul.circle li ul,ul.disc li ul{list-style:inherit}
ul.square{list-style-type:square}
ul.circle{list-style-type:circle}
ul.disc{list-style-type:disc}
ol li ul,ol li ol{margin-left:1.25em;margin-bottom:0}
dl dt{margin-bottom:.3125em;font-weight:bold}
dl dd{margin-bottom:1.25em}
abbr,acronym{text-transform:uppercase;font-size:90%;color:rgba(0,0,0,.8);border-bottom:1px dotted #ddd;cursor:help}
abbr{text-transform:none}
blockquote{margin:0 0 1.25em;padding:.5625em 1.25em 0 1.1875em;border-left:1px solid #ddd}
blockquote cite{display:block;font-size:.9375em;color:rgba(0,0,0,.6)}
blockquote cite::before{content:"\2014 \0020"}
blockquote cite a,blockquote cite a:visited{color:rgba(0,0,0,.6)}
blockquote,blockquote p{line-height:1.6;color:rgba(0,0,0,.85)}
@media screen and (min-width:768px){h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2}
h1{font-size:2.75em}
h2{font-size:2.3125em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.6875em}
h4{font-size:1.4375em}}
table{background:#fff;margin-bottom:1.25em;border:solid 1px #dedede;word-wrap:normal}
table thead,table tfoot{background:#f7f8f7}
table thead tr th,table thead tr td,table tfoot tr th,table tfoot tr td{padding:.5em .625em .625em;font-size:inherit;color:rgba(0,0,0,.8);text-align:left}
table tr th,table tr td{padding:.5625em .625em;font-size:inherit;color:rgba(0,0,0,.8)}
table tr.even,table tr.alt{background:#f8f8f7}
table thead tr th,table tfoot tr th,table tbody tr td,table tr td,table tfoot tr td{line-height:1.6}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2;word-spacing:-.05em}
h1 strong,h2 strong,h3 strong,#toctitle strong,.sidebarblock>.content>.title strong,h4 strong,h5 strong,h6 strong{font-weight:400}
.center{margin-left:auto;margin-right:auto}
.stretch{width:100%}
.clearfix::before,.clearfix::after,.float-group::before,.float-group::after{content:" ";display:table}
.clearfix::after,.float-group::after{clear:both}
:not(pre).nobreak{word-wrap:normal}
:not(pre).nowrap{white-space:nowrap}
:not(pre).pre-wrap{white-space:pre-wrap}
:not(pre):not([class^=L])>code{font-size:.9375em;font-style:normal!important;letter-spacing:0;padding:.1em .5ex;word-spacing:-.15em;background:#f7f7f8;-webkit-border-radius:4px;border-radius:4px;line-height:1.45;text-rendering:optimizeSpeed}
pre{color:rgba(0,0,0,.9);font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;line-height:1.45;text-rendering:optimizeSpeed}
pre code,pre pre{color:inherit;font-size:inherit;line-height:inherit}
pre>code{display:block}
pre.nowrap,pre.nowrap pre{white-space:pre;word-wrap:normal}
em em{font-style:normal}
strong strong{font-weight:400}
.keyseq{color:rgba(51,51,51,.8)}
kbd{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;display:inline-block;color:rgba(0,0,0,.8);font-size:.65em;line-height:1.45;background:#f7f7f7;border:1px solid #ccc;-webkit-border-radius:3px;border-radius:3px;-webkit-box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em white inset;box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em #fff inset;margin:0 .15em;padding:.2em .5em;vertical-align:middle;position:relative;top:-.1em;white-space:nowrap}
.keyseq kbd:first-child{margin-left:0}
.keyseq kbd:last-child{margin-right:0}
.menuseq,.menuref{color:#000}
.menuseq b:not(.caret),.menuref{font-weight:inherit}
.menuseq{word-spacing:-.02em}
.menuseq b.caret{font-size:1.25em;line-height:.8}
.menuseq i.caret{font-weight:bold;text-align:center;width:.45em}
b.button::before,b.button::after{position:relative;top:-1px;font-weight:400}
b.button::before{content:"[";padding:0 3px 0 2px}
b.button::after{content:"]";padding:0 2px 0 3px}
p a>code:hover{color:rgba(0,0,0,.9)}
#header,#content,#footnotes,#footer{width:100%;margin-left:auto;margin-right:auto;margin-top:0;margin-bottom:0;max-width:62.5em;*zoom:1;position:relative;padding-left:.9375em;padding-right:.9375em}
#header::before,#header::after,#content::before,#content::after,#footnotes::before,#footnotes::after,#footer::before,#footer::after{content:" ";display:table}
#header::after,#content::after,#footnotes::after,#footer::after{clear:both}
#content{margin-top:1.25em}
#content::before{content:none}
#header>h1:first-child{color:rgba(0,0,0,.85);margin-top:2.25rem;margin-bottom:0}
#header>h1:first-child+#toc{margin-top:8px;border-top:1px solid #dddddf}
#header>h1:only-child,body.toc2 #header>h1:nth-last-child(2){border-bottom:1px solid #dddddf;padding-bottom:8px}
#header .details{border-bottom:1px solid #dddddf;line-height:1.45;padding-top:.25em;padding-bottom:.25em;padding-left:.25em;color:rgba(0,0,0,.6);display:-ms-flexbox;display:-webkit-flex;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap}
#header .details span:first-child{margin-left:-.125em}
#header .details span.email a{color:rgba(0,0,0,.85)}
#header .details br{display:none}
#header .details br+span::before{content:"\00a0\2013\00a0"}
#header .details br+span.author::before{content:"\00a0\22c5\00a0";color:rgba(0,0,0,.85)}
#header .details br+span#revremark::before{content:"\00a0|\00a0"}
#header #revnumber{text-transform:capitalize}
#header #revnumber::after{content:"\00a0"}
#content>h1:first-child:not([class]){color:rgba(0,0,0,.85);border-bottom:1px solid #dddddf;padding-bottom:8px;margin-top:0;padding-top:1rem;margin-bottom:1.25rem}
#toc{border-bottom:1px solid #e7e7e9;padding-bottom:.5em}
#toc>ul{margin-left:.125em}
#toc ul.sectlevel0>li>a{font-style:italic}
#toc ul.sectlevel0 ul.sectlevel1{margin:.5em 0}
#toc ul{font-family:"Open Sans","DejaVu Sans",sans-serif;list-style-type:none}
#toc li{line-height:1.3334;margin-top:.3334em}
#toc a{text-decoration:none}
#toc a:active{text-decoration:underline}
#toctitle{color:#7a2518;font-size:1.2em}
@media screen and (min-width:768px){#toctitle{font-size:1.375em}
body.toc2{padding-left:15em;padding-right:0}
#toc.toc2{margin-top:0!important;background:#f8f8f7;position:fixed;width:15em;left:0;top:0;border-right:1px solid #e7e7e9;border-top-width:0!important;border-bottom-width:0!important;z-index:1000;padding:1.25em 1em;height:100%;overflow:auto}
#toc.toc2 #toctitle{margin-top:0;margin-bottom:.8rem;font-size:1.2em}
#toc.toc2>ul{font-size:.9em;margin-bottom:0}
#toc.toc2 ul ul{margin-left:0;padding-left:1em}
#toc.toc2 ul.sectlevel0 ul.sectlevel1{padding-left:0;margin-top:.5em;margin-bottom:.5em}
body.toc2.toc-right{padding-left:0;padding-right:15em}
body.toc2.toc-right #toc.toc2{border-right-width:0;border-left:1px solid #e7e7e9;left:auto;right:0}}
@media screen and (min-width:1280px){body.toc2{padding-left:20em;padding-right:0}
#toc.toc2{width:20em}
#toc.toc2 #toctitle{font-size:1.375em}
#toc.toc2>ul{font-size:.95em}
#toc.toc2 ul ul{padding-left:1.25em}
body.toc2.toc-right{padding-left:0;padding-right:20em}}
#content #toc{border-style:solid;border-width:1px;border-color:#e0e0dc;margin-bottom:1.25em;padding:1.25em;background:#f8f8f7;-webkit-border-radius:4px;border-radius:4px}
#content #toc>:first-child{margin-top:0}
#content #toc>:last-child{margin-bottom:0}
#footer{max-width:none;background:rgba(0,0,0,.8);padding:1.25em}
#footer-text{color:rgba(255,255,255,.8);line-height:1.44}
#content{margin-bottom:.625em}
.sect1{padding-bottom:.625em}
@media screen and (min-width:768px){#content{margin-bottom:1.25em}
.sect1{padding-bottom:1.25em}}
.sect1:last-child{padding-bottom:0}
.sect1+.sect1{border-top:1px solid #e7e7e9}
#content h1>a.anchor,h2>a.anchor,h3>a.anchor,#toctitle>a.anchor,.sidebarblock>.content>.title>a.anchor,h4>a.anchor,h5>a.anchor,h6>a.anchor{position:absolute;z-index:1001;width:1.5ex;margin-left:-1.5ex;display:block;text-decoration:none!important;visibility:hidden;text-align:center;font-weight:400}
#content h1>a.anchor::before,h2>a.anchor::before,h3>a.anchor::before,#toctitle>a.anchor::before,.sidebarblock>.content>.title>a.anchor::before,h4>a.anchor::before,h5>a.anchor::before,h6>a.anchor::before{content:"\00A7";font-size:.85em;display:block;padding-top:.1em}
#content h1:hover>a.anchor,#content h1>a.anchor:hover,h2:hover>a.anchor,h2>a.anchor:hover,h3:hover>a.anchor,#toctitle:hover>a.anchor,.sidebarblock>.content>.title:hover>a.anchor,h3>a.anchor:hover,#toctitle>a.anchor:hover,.sidebarblock>.content>.title>a.anchor:hover,h4:hover>a.anchor,h4>a.anchor:hover,h5:hover>a.anchor,h5>a.anchor:hover,h6:hover>a.anchor,h6>a.anchor:hover{visibility:visible}
#content h1>a.link,h2>a.link,h3>a.link,#toctitle>a.link,.sidebarblock>.content>.title>a.link,h4>a.link,h5>a.link,h6>a.link{color:#ba3925;text-decoration:none}
#content h1>a.link:hover,h2>a.link:hover,h3>a.link:hover,#toctitle>a.link:hover,.sidebarblock>.content>.title>a.link:hover,h4>a.link:hover,h5>a.link:hover,h6>a.link:hover{color:#a53221}
details,.audioblock,.imageblock,.literalblock,.listingblock,.stemblock,.videoblock{margin-bottom:1.25em}
details>summary:first-of-type{cursor:pointer;display:list-item;outline:none;margin-bottom:.75em}
.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{text-rendering:optimizeLegibility;text-align:left;font-family:"Noto Serif","DejaVu Serif",serif;font-size:1rem;font-style:italic}
table.tableblock.fit-content>caption.title{white-space:nowrap;width:0}
.paragraph.lead>p,#preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:1.21875em;line-height:1.6;color:rgba(0,0,0,.85)}
table.tableblock #preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:inherit}
.admonitionblock>table{border-collapse:separate;border:0;background:none;width:100%}
.admonitionblock>table td.icon{text-align:center;width:80px}
.admonitionblock>table td.icon img{max-width:none}
.admonitionblock>table td.icon .title{font-weight:bold;font-family:"Open Sans","DejaVu Sans",sans-serif;text-transform:uppercase}
.admonitionblock>table td.content{padding-left:1.125em;padding-right:1.25em;border-left:1px solid #dddddf;color:rgba(0,0,0,.6);word-wrap:anywhere}
.admonitionblock>table td.content>:last-child>:last-child{margin-bottom:0}
.exampleblock>.content{border-style:solid;border-width:1px;border-color:#e6e6e6;margin-bottom:1.25em;padding:1.25em;background:#fff;-webkit-border-radius:4px;border-radius:4px}
.exampleblock>.content>:first-child{margin-top:0}
.exampleblock>.content>:last-child{margin-bottom:0}
.sidebarblock{border-style:solid;border-width:1px;border-color:#dbdbd6;margin-bottom:1.25em;padding:1.25em;background:#f3f3f2;-webkit-border-radius:4px;border-radius:4px}
.sidebarblock>:first-child{margin-top:0}
.sidebarblock>:last-child{margin-bottom:0}
.sidebarblock>.content>.title{color:#7a2518;margin-top:0;text-align:center}
.exampleblock>.content>:last-child>:last-child,.exampleblock>.content .olist>ol>li:last-child>:last-child,.exampleblock>.content .ulist>ul>li:last-child>:last-child,.exampleblock>.content .qlist>ol>li:last-child>:last-child,.sidebarblock>.content>:last-child>:last-child,.sidebarblock>.content .olist>ol>li:last-child>:last-child,.sidebarblock>.content .ulist>ul>li:last-child>:last-child,.sidebarblock>.content .qlist>ol>li:last-child>:last-child{margin-bottom:0}
.literalblock pre,.listingblock>.content>pre{-webkit-border-radius:4px;border-radius:4px;overflow-x:auto;padding:1em;font-size:.8125em}
@media screen and (min-width:768px){.literalblock pre,.listingblock>.content>pre{font-size:.90625em}}
@media screen and (min-width:1280px){.literalblock pre,.listingblock>.content>pre{font-size:1em}}
.literalblock pre,.listingblock>.content>pre:not(.highlight),.listingblock>.content>pre[class="highlight"],.listingblock>.content>pre[class^="highlight "]{background:#f7f7f8}
.literalblock.output pre{color:#f7f7f8;background:rgba(0,0,0,.9)}
.listingblock>.content{position:relative}
.listingblock code[data-lang]::before{display:none;content:attr(data-lang);position:absolute;font-size:.75em;top:.425rem;right:.5rem;line-height:1;text-transform:uppercase;color:inherit;opacity:.5}
.listingblock:hover code[data-lang]::before{display:block}
.listingblock.terminal pre .command::before{content:attr(data-prompt);padding-right:.5em;color:inherit;opacity:.5}
.listingblock.terminal pre .command:not([data-prompt])::before{content:"$"}
.listingblock pre.highlightjs{padding:0}
.listingblock pre.highlightjs>code{padding:1em;-webkit-border-radius:4px;border-radius:4px}
.listingblock pre.prettyprint{border-width:0}
.prettyprint{background:#f7f7f8}
pre.prettyprint .linenums{line-height:1.45;margin-left:2em}
pre.prettyprint li{background:none;list-style-type:inherit;padding-left:0}
pre.prettyprint li code[data-lang]::before{opacity:1}
pre.prettyprint li:not(:first-child) code[data-lang]::before{display:none}
table.linenotable{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.linenotable td[class]{color:inherit;vertical-align:top;padding:0;line-height:inherit;white-space:normal}
table.linenotable td.code{padding-left:.75em}
table.linenotable td.linenos{border-right:1px solid currentColor;opacity:.35;padding-right:.5em}
pre.pygments .lineno{border-right:1px solid currentColor;opacity:.35;display:inline-block;margin-right:.75em}
pre.pygments .lineno::before{content:"";margin-right:-.125em}
.quoteblock{margin:0 1em 1.25em 1.5em;display:table}
.quoteblock:not(.excerpt)>.title{margin-left:-1.5em;margin-bottom:.75em}
.quoteblock blockquote,.quoteblock p{color:rgba(0,0,0,.85);font-size:1.15rem;line-height:1.75;word-spacing:.1em;letter-spacing:0;font-style:italic;text-align:justify}
.quoteblock blockquote{margin:0;padding:0;border:0}
.quoteblock blockquote::before{content:"\201c";float:left;font-size:2.75em;font-weight:bold;line-height:.6em;margin-left:-.6em;color:#7a2518;text-shadow:0 1px 2px rgba(0,0,0,.1)}
.quoteblock blockquote>.paragraph:last-child p{margin-bottom:0}
.quoteblock .attribution{margin-top:.75em;margin-right:.5ex;text-align:right}
.verseblock{margin:0 1em 1.25em}
.verseblock pre{font-family:"Open Sans","DejaVu Sans",sans;font-size:1.15rem;color:rgba(0,0,0,.85);font-weight:300;text-rendering:optimizeLegibility}
.verseblock pre strong{font-weight:400}
.verseblock .attribution{margin-top:1.25rem;margin-left:.5ex}
.quoteblock .attribution,.verseblock .attribution{font-size:.9375em;line-height:1.45;font-style:italic}
.quoteblock .attribution br,.verseblock .attribution br{display:none}
.quoteblock .attribution cite,.verseblock .attribution cite{display:block;letter-spacing:-.025em;color:rgba(0,0,0,.6)}
.quoteblock.abstract blockquote::before,.quoteblock.excerpt blockquote::before,.quoteblock .quoteblock blockquote::before{display:none}
.quoteblock.abstract blockquote,.quoteblock.abstract p,.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{line-height:1.6;word-spacing:0}
.quoteblock.abstract{margin:0 1em 1.25em;display:block}
.quoteblock.abstract>.title{margin:0 0 .375em;font-size:1.15em;text-align:center}
.quoteblock.excerpt>blockquote,.quoteblock .quoteblock{padding:0 0 .25em 1em;border-left:.25em solid #dddddf}
.quoteblock.excerpt,.quoteblock .quoteblock{margin-left:0}
.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{color:inherit;font-size:1.0625rem}
.quoteblock.excerpt .attribution,.quoteblock .quoteblock .attribution{color:inherit;text-align:left;margin-right:0}
p.tableblock:last-child{margin-bottom:0}
td.tableblock>.content{margin-bottom:1.25em;word-wrap:anywhere}
td.tableblock>.content>:last-child{margin-bottom:-1.25em}
table.tableblock,th.tableblock,td.tableblock{border:0 solid #dedede}
table.grid-all>*>tr>*{border-width:1px}
table.grid-cols>*>tr>*{border-width:0 1px}
table.grid-rows>*>tr>*{border-width:1px 0}
table.frame-all{border-width:1px}
table.frame-ends{border-width:1px 0}
table.frame-sides{border-width:0 1px}
table.frame-none>colgroup+*>:first-child>*,table.frame-sides>colgroup+*>:first-child>*{border-top-width:0}
table.frame-none>:last-child>:last-child>*,table.frame-sides>:last-child>:last-child>*{border-bottom-width:0}
table.frame-none>*>tr>:first-child,table.frame-ends>*>tr>:first-child{border-left-width:0}
table.frame-none>*>tr>:last-child,table.frame-ends>*>tr>:last-child{border-right-width:0}
table.stripes-all tr,table.stripes-odd tr:nth-of-type(odd),table.stripes-even tr:nth-of-type(even),table.stripes-hover tr:hover{background:#f8f8f7}
th.halign-left,td.halign-left{text-align:left}
th.halign-right,td.halign-right{text-align:right}
th.halign-center,td.halign-center{text-align:center}
th.valign-top,td.valign-top{vertical-align:top}
th.valign-bottom,td.valign-bottom{vertical-align:bottom}
th.valign-middle,td.valign-middle{vertical-align:middle}
table thead th,table tfoot th{font-weight:bold}
tbody tr th{background:#f7f8f7}
tbody tr th,tbody tr th p,tfoot tr th,tfoot tr th p{color:rgba(0,0,0,.8);font-weight:bold}
p.tableblock>code:only-child{background:none;padding:0}
p.tableblock{font-size:1em}
ol{margin-left:1.75em}
ul li ol{margin-left:1.5em}
dl dd{margin-left:1.125em}
dl dd:last-child,dl dd:last-child>:last-child{margin-bottom:0}
ol>li p,ul>li p,ul dd,ol dd,.olist .olist,.ulist .ulist,.ulist .olist,.olist .ulist{margin-bottom:.625em}
ul.checklist,ul.none,ol.none,ul.no-bullet,ol.no-bullet,ol.unnumbered,ul.unstyled,ol.unstyled{list-style-type:none}
ul.no-bullet,ol.no-bullet,ol.unnumbered{margin-left:.625em}
ul.unstyled,ol.unstyled{margin-left:0}
ul.checklist{margin-left:.625em}
ul.checklist li>p:first-child>.fa-square-o:first-child,ul.checklist li>p:first-child>.fa-check-square-o:first-child{width:1.25em;font-size:.8em;position:relative;bottom:.125em}
ul.checklist li>p:first-child>input[type="checkbox"]:first-child{margin-right:.25em}
ul.inline{display:-ms-flexbox;display:-webkit-box;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap;list-style:none;margin:0 0 .625em -1.25em}
ul.inline>li{margin-left:1.25em}
.unstyled dl dt{font-weight:400;font-style:normal}
ol.arabic{list-style-type:decimal}
ol.decimal{list-style-type:decimal-leading-zero}
ol.loweralpha{list-style-type:lower-alpha}
ol.upperalpha{list-style-type:upper-alpha}
ol.lowerroman{list-style-type:lower-roman}
ol.upperroman{list-style-type:upper-roman}
ol.lowergreek{list-style-type:lower-greek}
.hdlist>table,.colist>table{border:0;background:none}
.hdlist>table>tbody>tr,.colist>table>tbody>tr{background:none}
td.hdlist1,td.hdlist2{vertical-align:top;padding:0 .625em}
td.hdlist1{font-weight:bold;padding-bottom:1.25em}
td.hdlist2{word-wrap:anywhere}
.literalblock+.colist,.listingblock+.colist{margin-top:-.5em}
.colist td:not([class]):first-child{padding:.4em .75em 0;line-height:1;vertical-align:top}
.colist td:not([class]):first-child img{max-width:none}
.colist td:not([class]):last-child{padding:.25em 0}
.thumb,.th{line-height:0;display:inline-block;border:solid 4px #fff;-webkit-box-shadow:0 0 0 1px #ddd;box-shadow:0 0 0 1px #ddd}
.imageblock.left{margin:.25em .625em 1.25em 0}
.imageblock.right{margin:.25em 0 1.25em .625em}
.imageblock>.title{margin-bottom:0}
.imageblock.thumb,.imageblock.th{border-width:6px}
.imageblock.thumb>.title,.imageblock.th>.title{padding:0 .125em}
.image.left,.image.right{margin-top:.25em;margin-bottom:.25em;display:inline-block;line-height:0}
.image.left{margin-right:.625em}
.image.right{margin-left:.625em}
a.image{text-decoration:none;display:inline-block}
a.image object{pointer-events:none}
sup.footnote,sup.footnoteref{font-size:.875em;position:static;vertical-align:super}
sup.footnote a,sup.footnoteref a{text-decoration:none}
sup.footnote a:active,sup.footnoteref a:active{text-decoration:underline}
#footnotes{padding-top:.75em;padding-bottom:.75em;margin-bottom:.625em}
#footnotes hr{width:20%;min-width:6.25em;margin:-.25em 0 .75em;border-width:1px 0 0}
#footnotes .footnote{padding:0 .375em 0 .225em;line-height:1.3334;font-size:.875em;margin-left:1.2em;margin-bottom:.2em}
#footnotes .footnote a:first-of-type{font-weight:bold;text-decoration:none;margin-left:-1.05em}
#footnotes .footnote:last-of-type{margin-bottom:0}
#content #footnotes{margin-top:-.625em;margin-bottom:0;padding:.75em 0}
.gist .file-data>table{border:0;background:#fff;width:100%;margin-bottom:0}
.gist .file-data>table td.line-data{width:99%}
div.unbreakable{page-break-inside:avoid}
.big{font-size:larger}
.small{font-size:smaller}
.underline{text-decoration:underline}
.overline{text-decoration:overline}
.line-through{text-decoration:line-through}
.aqua{color:#00bfbf}
.aqua-background{background:#00fafa}
.black{color:#000}
.black-background{background:#000}
.blue{color:#0000bf}
.blue-background{background:#0000fa}
.fuchsia{color:#bf00bf}
.fuchsia-background{background:#fa00fa}
.gray{color:#606060}
.gray-background{background:#7d7d7d}
.green{color:#006000}
.green-background{background:#007d00}
.lime{color:#00bf00}
.lime-background{background:#00fa00}
.maroon{color:#600000}
.maroon-background{background:#7d0000}
.navy{color:#000060}
.navy-background{background:#00007d}
.olive{color:#606000}
.olive-background{background:#7d7d00}
.purple{color:#600060}
.purple-background{background:#7d007d}
.red{color:#bf0000}
.red-background{background:#fa0000}
.silver{color:#909090}
.silver-background{background:#bcbcbc}
.teal{color:#006060}
.teal-background{background:#007d7d}
.white{color:#bfbfbf}
.white-background{background:#fafafa}
.yellow{color:#bfbf00}
.yellow-background{background:#fafa00}
span.icon>.fa{cursor:default}
a span.icon>.fa{cursor:inherit}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note::before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip::before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning::before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution::before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important::before{content:"\f06a";color:#bf0000}
.conum[data-value]{display:inline-block;color:#fff!important;background:rgba(0,0,0,.8);-webkit-border-radius:50%;border-radius:50%;text-align:center;font-size:.75em;width:1.67em;height:1.67em;line-height:1.67em;font-family:"Open Sans","DejaVu Sans",sans-serif;font-style:normal;font-weight:bold}
.conum[data-value] *{color:#fff!important}
.conum[data-value]+b{display:none}
.conum[data-value]::after{content:attr(data-value)}
pre .conum[data-value]{position:relative;top:-.125em}
b.conum *{color:inherit!important}
.conum:not([data-value]):empty{display:none}
dt,th.tableblock,td.content,div.footnote{text-rendering:optimizeLegibility}
h1,h2,p,td.content,span.alt{letter-spacing:-.01em}
p strong,td.content strong,div.footnote strong{letter-spacing:-.005em}
p,blockquote,dt,td.content,span.alt{font-size:1.0625rem}
p{margin-bottom:1.25rem}
.sidebarblock p,.sidebarblock dt,.sidebarblock td.content,p.tableblock{font-size:1em}
.exampleblock>.content{background:#fffef7;border-color:#e0e0dc;-webkit-box-shadow:0 1px 4px #e0e0dc;box-shadow:0 1px 4px #e0e0dc}
.print-only{display:none!important}
@page{margin:1.25cm .75cm}
@media print{*{-webkit-box-shadow:none!important;box-shadow:none!important;text-shadow:none!important}
html{font-size:80%}
a{color:inherit!important;text-decoration:underline!important}
a.bare,a[href^="#"],a[href^="mailto:"]{text-decoration:none!important}
a[href^="http:"]:not(.bare)::after,a[href^="https:"]:not(.bare)::after{content:"(" attr(href) ")";display:inline-block;font-size:.875em;padding-left:.25em}
abbr[title]::after{content:" (" attr(title) ")"}
pre,blockquote,tr,img,object,svg{page-break-inside:avoid}
thead{display:table-header-group}
svg{max-width:100%}
p,blockquote,dt,td.content{font-size:1em;orphans:3;widows:3}
h2,h3,#toctitle,.sidebarblock>.content>.title{page-break-after:avoid}
#header,#content,#footnotes,#footer{max-width:none}
#toc,.sidebarblock,.exampleblock>.content{background:none!important}
#toc{border-bottom:1px solid #dddddf!important;padding-bottom:0!important}
body.book #header{text-align:center}
body.book #header>h1:first-child{border:0!important;margin:2.5em 0 1em}
body.book #header .details{border:0!important;display:block;padding:0!important}
body.book #header .details span:first-child{margin-left:0!important}
body.book #header .details br{display:block}
body.book #header .details br+span::before{content:none!important}
body.book #toc{border:0!important;text-align:left!important;padding:0!important;margin:0!important}
body.book #toc,body.book #preamble,body.book h1.sect0,body.book .sect1>h2{page-break-before:always}
.listingblock code[data-lang]::before{display:block}
#footer{padding:0 .9375em}
.hide-on-print{display:none!important}
.print-only{display:block!important}
.hide-for-print{display:none!important}
.show-for-print{display:inherit!important}}
@media print,amzn-kf8{#header>h1:first-child{margin-top:1.25rem}
.sect1{padding:0!important}
.sect1+.sect1{border:0}
#footer{background:none}
#footer-text{color:rgba(0,0,0,.6);font-size:.9em}}
@media amzn-kf8{#header,#content,#footnotes,#footer{padding:0}}
</style>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<style>
pre.rouge table td { padding: 5px; }
pre.rouge table pre { margin: 0; }
pre.rouge .cm {
  color: #999988;
  font-style: italic;
}
pre.rouge .cp {
  color: #999999;
  font-weight: bold;
}
pre.rouge .c1 {
  color: #999988;
  font-style: italic;
}
pre.rouge .cs {
  color: #999999;
  font-weight: bold;
  font-style: italic;
}
pre.rouge .c, pre.rouge .ch, pre.rouge .cd, pre.rouge .cpf {
  color: #999988;
  font-style: italic;
}
pre.rouge .err {
  color: #a61717;
  background-color: #e3d2d2;
}
pre.rouge .gd {
  color: #000000;
  background-color: #ffdddd;
}
pre.rouge .ge {
  color: #000000;
  font-style: italic;
}
pre.rouge .gr {
  color: #aa0000;
}
pre.rouge .gh {
  color: #999999;
}
pre.rouge .gi {
  color: #000000;
  background-color: #ddffdd;
}
pre.rouge .go {
  color: #888888;
}
pre.rouge .gp {
  color: #555555;
}
pre.rouge .gs {
  font-weight: bold;
}
pre.rouge .gu {
  color: #aaaaaa;
}
pre.rouge .gt {
  color: #aa0000;
}
pre.rouge .kc {
  color: #000000;
  font-weight: bold;
}
pre.rouge .kd {
  color: #000000;
  font-weight: bold;
}
pre.rouge .kn {
  color: #000000;
  font-weight: bold;
}
pre.rouge .kp {
  color: #000000;
  font-weight: bold;
}
pre.rouge .kr {
  color: #000000;
  font-weight: bold;
}
pre.rouge .kt {
  color: #445588;
  font-weight: bold;
}
pre.rouge .k, pre.rouge .kv {
  color: #000000;
  font-weight: bold;
}
pre.rouge .mf {
  color: #009999;
}
pre.rouge .mh {
  color: #009999;
}
pre.rouge .il {
  color: #009999;
}
pre.rouge .mi {
  color: #009999;
}
pre.rouge .mo {
  color: #009999;
}
pre.rouge .m, pre.rouge .mb, pre.rouge .mx {
  color: #009999;
}
pre.rouge .sa {
  color: #000000;
  font-weight: bold;
}
pre.rouge .sb {
  color: #d14;
}
pre.rouge .sc {
  color: #d14;
}
pre.rouge .sd {
  color: #d14;
}
pre.rouge .s2 {
  color: #d14;
}
pre.rouge .se {
  color: #d14;
}
pre.rouge .sh {
  color: #d14;
}
pre.rouge .si {
  color: #d14;
}
pre.rouge .sx {
  color: #d14;
}
pre.rouge .sr {
  color: #009926;
}
pre.rouge .s1 {
  color: #d14;
}
pre.rouge .ss {
  color: #990073;
}
pre.rouge .s, pre.rouge .dl {
  color: #d14;
}
pre.rouge .na {
  color: #008080;
}
pre.rouge .bp {
  color: #999999;
}
pre.rouge .nb {
  color: #0086B3;
}
pre.rouge .nc {
  color: #445588;
  font-weight: bold;
}
pre.rouge .no {
  color: #008080;
}
pre.rouge .nd {
  color: #3c5d5d;
  font-weight: bold;
}
pre.rouge .ni {
  color: #800080;
}
pre.rouge .ne {
  color: #990000;
  font-weight: bold;
}
pre.rouge .nf, pre.rouge .fm {
  color: #990000;
  font-weight: bold;
}
pre.rouge .nl {
  color: #990000;
  font-weight: bold;
}
pre.rouge .nn {
  color: #555555;
}
pre.rouge .nt {
  color: #000080;
}
pre.rouge .vc {
  color: #008080;
}
pre.rouge .vg {
  color: #008080;
}
pre.rouge .vi {
  color: #008080;
}
pre.rouge .nv, pre.rouge .vm {
  color: #008080;
}
pre.rouge .ow {
  color: #000000;
  font-weight: bold;
}
pre.rouge .o {
  color: #000000;
  font-weight: bold;
}
pre.rouge .w {
  color: #bbbbbb;
}
pre.rouge {
  background-color: #f8f8f8;
}
</style>
</head>
<body class="article toc2 toc-left">
<div id="header">
<h1>Mapping traits in biparental populations with genotyping-by-sequencing data</h1>
<div id="toc" class="toc2">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel1">
<li><a href="#_open_a_unix_shell">1. Open a UNIX shell</a></li>
<li><a href="#_set_up_your_analysis_directory">2. Set up your analysis directory</a></li>
<li><a href="#_read_mapping">3. Read mapping</a></li>
<li><a href="#_variant_calling_and_filtration">4. Variant calling and filtration</a></li>
<li><a href="#_import_the_genetic_variant_matrix_into_r">5. Import the genetic variant matrix into R</a></li>
<li><a href="#_quality_control_and_linkage_map_construction">6. Quality control and linkage map construction</a></li>
<li><a href="#_trait_mapping">7. Trait mapping</a></li>
</ul>
</div>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<link rel="stylesheet"  href="http://cdnjs.cloudflare.com/ajax/libs/font-awesome/3.1.0/css/font-awesome.min.css">
<div class="imageblock text-center">
<div class="content">
<img src="GBStutorial_flowchart.jpeg" alt="GBStutorial flowchart">
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_open_a_unix_shell">1. Open a UNIX shell</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Connect with <a href="https://en.wikipedia.org/wiki/PuTTY">PuTTY</a> to the server slurm-01.ipk-gatersleben.de.</p>
</li>
<li>
<p>Obtain a Kerberos ticket to access our filer resources.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">kinit
klist <i class="conum" data-value="1"></i><b>(1)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>A ticket is valid for seven days.</td>
</tr>
</table>
</div>
</li>
<li>
<p>You will see the prompt of the UNIX command-line interpreter, the <a href="https://en.wikipedia.org/wiki/Unix_shell">shell</a>.
It is called "shell" because it executes commands that do not belong to the operating system, the "kernel".</p>
</li>
<li>
<p>Check out your home directory.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">pwd
ls</span></code></pre>
</div>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_set_up_your_analysis_directory">2. Set up your analysis directory</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Let&#8217;s have a look at the GBS data.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">cd</span> /filer/projekte/gbs_course
<span class="nb">cd </span>data
<span class="nb">ls
ls</span> <span class="nt">-lhS</span> gbs_reads | less <i class="conum" data-value="1"></i><b>(1)</b>
man <span class="nb">ls
cd </span>gbs_reads
<span class="nb">ls</span> | <span class="nb">wc</span> <span class="nt">-l</span>
zcat Morex.fastq.gz | less <span class="nt">-S</span> <i class="conum" data-value="2"></i><b>(2)</b>
<span class="nb">cd</span> ..
<span class="nb">du</span> <span class="nt">-ch</span> gbs_reads</code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>To close <code>less</code>, press <code>q</code>.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The reads are in <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ format</a>.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Take a look at the phenotypes. There are two. One is binary: lateral spikelet fertility (commonly referred to in barley as row type). The other is quantitative (plant height).</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">less MxB_92_RILs_phenotypes.csv</code></pre>
</div>
</div>
</li>
<li>
<p>Create your working directory and change into it.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">cd</span> /filer/projekte/gbs_course
<span class="nb">mkdir</span> <span class="nt">-p</span> mascher/gbs <i class="conum" data-value="1"></i><b>(1)</b>
<span class="nb">cd </span>mascher/gbs</code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Change <code>mascher</code> to your name.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Create symbolic links to the GBS reads and the phenotype data.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">pwd <i class="conum" data-value="1"></i><b>(1)</b>

mkdir </span>gbs_reads
<span class="nv">datadir</span><span class="o">=</span><span class="s2">"/filer/projekte/gbs_course/data/"</span>
<span class="nb">ln</span> <span class="nt">-s</span> <span class="nv">$datadir</span>/gbs_reads/<span class="k">*</span>fastq.gz gbs_reads
<span class="nb">ls</span> <span class="nt">-l</span> gbs_reads | less <span class="nt">-S</span>
<span class="nb">ls </span>gbs_reads | <span class="nb">wc</span> <span class="nt">-l</span>

<span class="nb">ln</span> <span class="nt">-s</span> <span class="nv">$datadir</span>/MxB_92_RILs_phenotypes.csv</code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Check that you are in your working directory (<code>/filer/projekte/gbs_course/USER/gbs</code>).</td>
</tr>
</table>
</div>
</li>
<li>
<p>Count the reads. We use a for-loop to process all samples at once.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">zcat</span><span class="w"> </span><span class="n">gbs_reads</span><span class="o">/</span><span class="n">Morex.fastq.gz</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">wc</span><span class="w"> </span><span class="o">-</span><span class="n">l</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">Morex.len</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">gbs_reads</span><span class="o">/*</span><span class="n">fastq.gz</span><span class="p">;</span><span class="w"> </span><span class="n">do</span><span class="w">
 </span><span class="n">name</span><span class="o">=</span><span class="n">`basename $i | cut -d . -f 1`</span><span class="w">
 </span><span class="n">count</span><span class="o">=</span><span class="n">`zcat $i | wc -l`</span><span class="w">
 </span><span class="n">echo</span><span class="w"> </span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">$</span><span class="n">count</span><span class="w">
</span><span class="n">done</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">fastq_line_counts.txt</span><span class="w">

</span><span class="n">cat</span><span class="w"> </span><span class="n">fastq_line_counts.txt</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">awk</span><span class="w"> </span><span class="s1">'{print $1,$2/4}'</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">sort</span><span class="w"> </span><span class="o">-</span><span class="n">nk</span><span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">fastq_read_counts.txt</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>We use <a href="https://www.tutorialspoint.com/awk/index.htm">AWK</a> to divide the second column by 4.</td>
</tr>
</table>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_read_mapping">3. Read mapping</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>We will use the tools <a href="https://github.com/lh3/bwa">BWA</a> and <a href="http://www.htslib.org/doc/samtools.html">samtools</a> for read mapping and processing of alignment records, respectively.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">module load bwa samtools
module list</code></pre>
</div>
</div>
</li>
<li>
<p>Load the BWA index into shared memory. This step can be skipped, but pre-loadeding the index makes mapping a lot faster.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nv">ref</span><span class="o">=</span><span class="s1">'/filer/projekte/gbs_course/data/reference/Hordeum_vulgare_MorexV3_assembly.fasta'</span> <i class="conum" data-value="1"></i><b>(1)</b>
bwa shm <span class="nv">$ref</span>
bwa shm <span class="nt">-l</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>We use the <a href="https://academic.oup.com/plcell/article/33/6/1888/6169005">MorexV3</a> reference genome sequence.</td>
</tr>
</table>
</div>
</li>
<li>
<p>We will first run a toy example with 10,000 reads of the Morex sample.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">zcat gbs_reads/Morex.fastq.gz | <span class="nb">head</span> <span class="nt">-n</span> 40000 <span class="o">&gt;</span> Morex_10k.fastq
bwa mem <span class="nv">$ref</span> Morex_10k.fastq <span class="o">&gt;</span> Morex_10k.sam <i class="conum" data-value="1"></i><b>(1)</b>
samtools <span class="nb">sort </span>Morex_10k.sam <span class="o">&gt;</span> Morex_10k.bam <i class="conum" data-value="1"></i><b>(1)</b>
samtools view Morex_10k.bam | less <span class="nt">-S</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Alignments are stored in the <a href="https://en.wikipedia.org/wiki/SAM_(file_format)">Sequence Alignment Map format</a> (SAM) or its binary equivalent BAM.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Combine mapping and sorting into one command.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">bwa mem <span class="nv">$ref</span> Morex_10k.fastq | samtools <span class="nb">sort</span> <span class="o">&gt;</span> Morex_10k.bam</code></pre>
</div>
</div>
</li>
<li>
<p>Open a new PuTTY session. Start <a href="https://www.hamvocke.com/blog/a-quick-and-easy-guide-to-tmux/">tmux</a> to keep your commands running in the background.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">k5reauth tmux <i class="conum" data-value="1"></i><b>(1)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>k5reauth is used to prolong the Kerberos ticket for as long as the tmux session is running.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Try out tmux functionality.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">tmux <span class="nb">ls
</span>tmux rename mapping
tmux <span class="nb">ls
</span>tmux detach
tmux <span class="nb">ls
</span>tmux attach <span class="nt">-t</span> mapping</code></pre>
</div>
</div>
</li>
<li>
<p>Map all samples.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nv">ref</span><span class="o">=</span><span class="s1">'/filer/projekte/gbs_course/data/reference/Hordeum_vulgare_MorexV3_assembly.fasta'</span>
<span class="k">for </span>i <span class="k">in </span>gbs_reads/<span class="k">*</span>fastq.gz<span class="p">;</span> <span class="k">do
 </span><span class="nv">name</span><span class="o">=</span><span class="sb">`</span><span class="nb">echo</span> <span class="nv">$i</span> | <span class="nb">cut</span> <span class="nt">-d</span> <span class="nb">.</span> <span class="nt">-f</span> 1<span class="sb">`</span> <i class="conum" data-value="1"></i><b>(1)</b>
 bwa mem <span class="nt">-t</span> 4 <span class="nv">$ref</span> <span class="nv">$i</span> | samtools <span class="nb">sort</span> <span class="o">&gt;</span> <span class="nv">$name</span>.bam
<span class="k">done </span>2&gt; bwa.err <i class="conum" data-value="2"></i><b>(2)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Strip the extension: Morex.fastq.gz become Morex.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>To detach the tmux session, press <kbd>Ctrl-B</kbd> followed by <kbd>D</kbd>.</td>
</tr>
</table>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title=":heavy_exclamation_mark:"></i>
</td>
<td class="content">
If you forget to start bwa inside a tmux session, there is no way to prevent your job from aborting when you shutdown your laptop. Also without k5reauth programs cannot access filer resource after a maximum of ten hours.
</td>
</tr>
</table>
</div>
</li>
<li>
<p>Open a new terminal. Look at your jobs in the table of processes (<code>top</code>).</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">find gbs_reads | <span class="nb">grep</span> <span class="nt">-c</span> <span class="s1">'bam$'</span> <i class="conum" data-value="1"></i><b>(1)</b>
top <span class="nt">-u</span> mascher <i class="conum" data-value="2"></i><b>(2)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Show the number of BAM files created so far.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Replace mascher with your username.</td>
</tr>
</table>
</div>
</li>
<li>
<p>When the mapping is done, calculate the mapping rates for all samples.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="k">for </span>i <span class="k">in </span>gbs_reads/<span class="k">*</span>bam<span class="p">;</span> <span class="k">do
 </span><span class="nv">name</span><span class="o">=</span><span class="sb">`</span><span class="nb">basename</span> <span class="nv">$i</span> | <span class="nb">cut</span> <span class="nt">-d</span> <span class="nb">.</span> <span class="nt">-f</span> 1<span class="sb">`</span>
 <span class="nv">count</span><span class="o">=</span><span class="sb">`</span>samtools view <span class="nt">-cq</span> 20 <span class="nv">$i</span><span class="sb">`</span>
 <span class="nb">echo</span> <span class="nv">$name</span> <span class="nv">$count</span>
<span class="k">done</span> <span class="o">&gt;</span> mapped_reads.txt</code></pre>
</div>
</div>
</li>
<li>
<p>Combine the raw read counts and the mapping rates into one table.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">sort </span>fastq_read_counts.txt <span class="o">&gt;</span> tmp1 <i class="conum" data-value="1"></i><b>(1)</b>
<span class="nb">sort </span>mapped_reads.txt <span class="o">&gt;</span> tmp2 <i class="conum" data-value="1"></i><b>(1)</b>

<span class="nb">join </span>tmp1 tmp2 | <span class="nb">awk</span> <span class="s1">'{print $0,$3/$2*100}'</span> | <span class="nb">sort</span> <span class="nt">-nk</span> 4 | column <span class="nt">-t</span> <span class="o">&gt;</span> mapping_stats.txt <i class="conum" data-value="2"></i><b>(2)</b>

<span class="nb">rm</span> <span class="nt">-f</span> tmp1 tmp2</code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>To combine two lists with <a href="https://linux.die.net/man/1/join">join</a>, both lists need to be sorted on the common ID column.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td><a href="https://linux.die.net/man/1/column">column</a> is used to align columns.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Get a list of all BAM files.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">ls </span>gbs_reads/<span class="k">*</span>bam | <span class="nb">sort</span> <span class="o">&gt;</span> bam_list.txt</code></pre>
</div>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_variant_calling_and_filtration">4. Variant calling and filtration</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Open a new tmux session and load <a href="https://samtools.github.io/bcftools/howtos/index.html">bcftools</a>.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">tmux <i class="conum" data-value="1"></i><b>(1)</b>
tmux rename variant_call
module load bcftools</code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The variant calling will run for some time, so run it inside <code>tmux</code>.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Run the variant calling.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nv">ref</span><span class="o">=</span><span class="s1">'/filer/projekte/gbs_course/data/reference/Hordeum_vulgare_MorexV3_assembly.fasta'</span>
<span class="c">#bamlist='bam_list.txt'</span>
<span class="nv">bamlist</span><span class="o">=</span><span class="s1">'/filer-dg/agruppen/dg6/mascher/DG/mxb_course_221201/try_221216/bam_list_Martin.txt'</span> <i class="conum" data-value="1"></i><b>(1)</b>

<span class="nv">vcf</span><span class="o">=</span><span class="s1">'bcftools_SNP_calling.vcf'</span> <i class="conum" data-value="2"></i><b>(2)</b>

bcftools mpileup <span class="nt">--skip-indels</span> <span class="nt">--fasta-ref</span> <span class="nv">$ref</span> <span class="nt">--bam-list</span> <span class="nv">$bamlist</span> <span class="nt">--min-MQ</span> 20 <span class="nt">--annotate</span> AD <span class="se">\</span><i class="conum" data-value="3"></i><b>(3)</b>
 | bcftools view <span class="nt">-i</span> <span class="s1">'INFO/DP &gt; 100'</span> <span class="se">\</span><i class="conum" data-value="4"></i><b>(4)</b> <i class="conum" data-value="5"></i><b>(5)</b>
 | bcftools call <span class="nt">--consensus-caller</span> <span class="nt">--variants-only</span> <span class="nt">--output</span> <span class="nv">$vcf</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>List of pre-computed BAM files.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Output file in <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">variant call format</a> (VCF). <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">Here</a> are the specifications of the VCF format.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>We ignore insertions and deletions (<code>--skip-indels</code>), consider only SNPs with a quality score no smaller than 20 (<code>--min-MQ 20</code>) and add allelic depth information (<code>--annotate AD</code>) for all genotype calls.</td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td>Remove sites with fewer than 100 reads across all samples. We are only interested in sites that have at least two supporting reads in nine tenths of the samples.</td>
</tr>
<tr>
<td><i class="conum" data-value="5"></i><b>5</b></td>
<td>The backslash \ character is used to split long commands across multiple lines.  When pasting the commands or editing them, make sure that no white space follows the backslash.  Otherwise, the shell will interpret the lines as belonging to different commands. Also multi-line commands do not tolerate intervening command line (starting the hash sign #).</td>
</tr>
</table>
</div>
</li>
<li>
<p>Filter the variant calls.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nv">filter</span><span class="o">=</span><span class="s1">'/filer/projekte/gbs_course/scripts/filter_vcf.zsh'</span>
<span class="c">#vcf='bcftools_SNP_calling.vcf'</span>
<span class="nv">vcf</span><span class="o">=</span><span class="s1">'/filer-dg/agruppen/dg6/mascher/DG/mxb_course_221201/try_221216/bcftools_SNP_calling.vcf'</span> <i class="conum" data-value="1"></i><b>(1)</b>
<span class="nv">fvcf</span><span class="o">=</span><span class="s1">'bcftools_SNP_calling_filtered.vcf'</span>

<span class="nv">$filter</span> <span class="nt">--vcf</span> <span class="nv">$vcf</span> <span class="nt">--dphom</span> 2 <span class="nt">--dphet</span> 4 <span class="nt">--minmaf</span> 0.2 <span class="nt">--minpresent</span> 0.9 <span class="nt">--minhomp</span> 0.9 <span class="o">&gt;</span> <span class="nv">$fvcf</span> <i class="conum" data-value="2"></i><b>(2)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Path to pre-computed VCF file.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>We keep homozygous genotype calls if they have at least two supporting reads; heterozygous calls are accepted if they are supported by no fewer than four reads. SNPs with a minor allele frequency below 20 % or less than 90 % present calls or less than 90 % homozyous calls are discarded.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Change the column names of the VCF files to match the row names in the phenotype table.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">less MxB_92_RILs_phenotypes.csv
bcftools query <span class="nt">-l</span> bcftools_SNP_calling_filtered.vcf | less <span class="nt">-S</span>
bcftools query <span class="nt">-l</span> <span class="nv">$fvcf</span> | <span class="nb">cut</span> <span class="nt">-d</span> / <span class="nt">-f</span> 2 | <span class="nb">cut</span> <span class="nt">-d</span> <span class="nb">.</span> <span class="nt">-f</span> 1 <span class="o">&gt;</span> new_sample_names.txt
bcftools reheader <span class="nt">-s</span> new_sample_names.txt <span class="nv">$fvcf</span> <span class="o">&gt;</span> bcftools_SNP_calling_filtered_newNames.vcf</code></pre>
</div>
</div>
</li>
<li>
<p>Review the VCF file.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh"><span class="nb">grep</span> <span class="nt">-v</span> <span class="s1">'^##'</span> bcftools_SNP_calling_filtered_newNames.vcf | column <span class="nt">-t</span> | less <span class="nt">-S</span></code></pre>
</div>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_import_the_genetic_variant_matrix_into_r">5. Import the genetic variant matrix into R</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Start R.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="sh">module load R/4.1.1
R</code></pre>
</div>
</div>
</li>
<li>
<p>R is a widely used programming language in data science. There are very many tutorials, e.g. <a href="https://www.statmethods.net/r-tutorial/index.html">this one</a>.</p>
</li>
<li>
<p>Load the required libraries.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">.libPaths</span><span class="p">(</span><span class="s2">"/filer/projekte/gbs_course/Rlibs/4.1.1"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>

</span><span class="n">library</span><span class="p">(</span><span class="s2">"qtl"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b>
</span><span class="n">library</span><span class="p">(</span><span class="s2">"ASMap"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="3"></i><b>(3)</b>
</span><span class="n">library</span><span class="p">(</span><span class="s2">"utl"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="4"></i><b>(4)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Set the path where the R libraries are located.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td><a href="https://rqtl.org">R/qtl</a> is package for QTL mapping. Several tutorials are available <a href="https://rqtl.org/tutorials/">here</a>.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td><a href="https://cran.r-project.org/web/packages/ASMap/ASMap.pdf">R/ASMap</a> is package for linkage map constuction. It implements the <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000212">MSTMAP</a> algorithm.
A detailed tutorial is available <a href="https://cran.r-project.org/web/packages/ASMap/vignettes/asmapvignette.pdf">here</a>.</td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td><a href="https://github.com/gact/utl">utl</a> provides utility functions for R/qtl, one of which we use to convert VCF to R/qtl format.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Convert the VCF to R/qtl format. Example files are found <a href="https://rqtl.org/sampledata/">here</a>.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">vcf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'bcftools_SNP_calling_filtered_newNames.vcf'</span><span class="w">
</span><span class="n">ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="s2">"new_sample_names.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">head</span><span class="o">=</span><span class="nb">F</span><span class="p">)[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w">
</span><span class="n">genfile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'bcftools_SNP_calling_geno.csv'</span><span class="w">
</span><span class="n">founders</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Morex"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Barke"</span><span class="p">)</span><span class="w">
</span><span class="n">samples</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">setdiff</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="w"> </span><span class="n">founders</span><span class="p">)</span><span class="w">

</span><span class="n">convert_vcf_to_genfile</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span><span class="w"> </span><span class="n">genfile</span><span class="p">,</span><span class="w"> </span><span class="n">samples</span><span class="p">,</span><span class="w"> </span><span class="n">founders</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>This function writes a text file in R/qtl&#8217;s "csvs" format to disk. The output filename is <code>genfile</code>.</td>
</tr>
</table>
</div>
</li>
<li>
<p>The conversion function does not take genomic coordinates into account, so markers are ordered correctly, but equidistant. We add a line to the CSV file to correct this.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">cmd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"sed -Ei '1{p; s/id|chr.H://g; h; d}; 2G'"</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">paste</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span><span class="w"> </span><span class="n">genfile</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">system</span><span class="p">()</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b> <i class="conum" data-value="3"></i><b>(3)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The <a href="https://www.gnu.org/software/sed/manual/sed.html#Introduction">sed</a> command retrieves the genomic coordinates from the first line and inserts them as a new line after line 2.
This could also be achieved with a text editor. The command, at the cost of arcanity, forgoes error-prone manual editing.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The <code>sed</code> command is called from inside R and modifies the file in place.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>Alternatively, you write this command as <code>system(paste(cmd, genfile))</code>. Before R version 4.0 (or so), this was the only way to do it. The pipe-like syntax is a recent addition to R.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Read the genotype and phenotype data into an R/qtl cross object.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pheno</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'MxB_92_RILs_phenotypes.csv'</span><span class="w">
</span><span class="n">read.cross</span><span class="p">(</span><span class="n">format</span><span class="o">=</span><span class="s2">"csvs"</span><span class="p">,</span><span class="w"> </span><span class="n">genfile</span><span class="o">=</span><span class="n">genfile</span><span class="p">,</span><span class="w"> </span><span class="n">phefile</span><span class="o">=</span><span class="n">pheno</span><span class="p">,</span><span class="w"> </span><span class="n">crosstype</span><span class="o">=</span><span class="w"> </span><span class="s2">"f2"</span><span class="p">,</span><span class="w">  </span><span class="n">genotypes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"AA"</span><span class="p">,</span><span class="s2">"AB"</span><span class="p">,</span><span class="s2">"BB"</span><span class="p">))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span><span class="w">
</span><span class="n">convert2bcsft</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">BC.gen</span><span class="o">=</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">F.gen</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="w">  </span><span class="n">estimate.map</span><span class="o">=</span><span class="nb">F</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">summary</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w">
</span><span class="n">summary.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>A population of recombinant inbred lines is read in as "f2". <code>convert2bcsft()</code> is used to set the correct number of selfing generations.  See <a href="https://cran.r-project.org/web/packages/qtl/vignettes/bcsft.pdf">the documentation</a> of that function.  We disable the <code>estimate.map</code> option because we will rely on R/ASMap to construct a genetic linkage map.</td>
</tr>
</table>
</div>
</li>
<li>
<p>R/qtl offers several functions to extract basic information from "cross" objects.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">nind</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="c1"># number of inviduals</span><span class="w">
</span><span class="n">nchr</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="c1"># number of chromosomes (actually linkage group)</span><span class="w">
</span><span class="n">totmar</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="c1"># total number of markers</span><span class="w">
</span><span class="n">nmar</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="c1"># number of markers per chromosome</span><span class="w">
</span><span class="n">nphe</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="c1"># number of phenotypes</span></code></pre>
</div>
</div>
</li>
<li>
<p>Plot a summary of the phenotypes. Copy the PDF file with <a href="https://winscp.net/eng/download.php">WinSCP</a> and take a look at it with the Acrobat Reader.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_pheno.pdf"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">plotPheno</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">pheno.col</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"phenotypes"</span><span class="p">)</span><span class="w">
</span><span class="n">plotPheno</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">pheno.col</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Open a PDF file to plot to. The file is created in the current working directory.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Close the file (switch <strong>off</strong> the plotting <strong>device</strong>). Don&#8217;t forget to call <code>dev.off()</code>. Otherwise the PDF file will be empty or invalid.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Change the scale of the linkage map from base pairs to megabase pairs and plot the distribution of markers along chromosomes.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">rescalemap</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">/</span><span class="m">1e6</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span><span class="w">
</span><span class="n">summary.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_map.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">plot.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"Physical map"</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s2">"Location (Mb)"</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
</li>
<li>
<p>Create a copy of the cross object with the physical map for later comparison.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">copy</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb_physical</span><span class="w">
</span><span class="n">saveRDS</span><span class="p">(</span><span class="n">mxb_physical</span><span class="p">,</span><span class="w"> </span><span class="s2">"mxb_physical.Rds"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">readRDS</span><span class="p">(</span><span class="s2">"mxb_physical.Rds"</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb_physical</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Any R object can be saved to and read from disk, respectively, with <code>saveRDS()</code> and <code>readRDS().</code></td>
</tr>
</table>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_quality_control_and_linkage_map_construction">6. Quality control and linkage map construction</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Karl Broman <a href="https://rqtl.org/tutorials/geneticmaps.pdf">pointed out</a> that a reference genome sequence
has obviated the need for linkage map construction in mouse;
the same applies to barley.
Although marker order is known, constructing a linkage map from scratch is a means of quality control.
If there are issues with the data, the genetic map will be off the mark.</p>
</li>
<li>
<p>A map off the cuff is not too bad. The only worry is that it&#8217;s about a fifth longer than expected.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">summary.map</span><span class="p">(</span><span class="n">mstmap</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">id</span><span class="o">=</span><span class="s2">"id"</span><span class="p">))</span></code></pre>
</div>
</div>
</li>
<li>
<p>Let&#8217;s see if we can do better. The usual suscepts are bad markers and odd individuals.</p>
</li>
<li>
<p>Remove duplicated markers. If there are groups of markers that differ only in their patterns of missing data, keep only one representative.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">findDupMarkers</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">exact.only</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">dups</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">unlist</span><span class="p">(</span><span class="n">dups</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">length</span><span class="p">()</span><span class="w">

</span><span class="n">mxb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">drop.markers</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">unlist</span><span class="p">(</span><span class="n">dups</span><span class="p">))</span><span class="w">
</span><span class="n">summary</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>exact.only=FALSE ignores differences in missingness.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Remove duplicated individuals.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">cg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">comparegeno</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"compare_geno.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">hist</span><span class="p">(</span><span class="n">cg</span><span class="p">[</span><span class="n">lower.tri</span><span class="p">(</span><span class="n">cg</span><span class="p">)],</span><span class="w"> </span><span class="n">breaks</span><span class="o">=</span><span class="n">seq</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">len</span><span class="o">=</span><span class="m">101</span><span class="p">),</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"No. matching genotypes"</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="n">cg</span><span class="p">,</span><span class="w"> </span><span class="n">thresh</span><span class="o">=</span><span class="m">0.9</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">dups</span><span class="w">
</span><span class="n">subset</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">ind</span><span class="o">=</span><span class="n">setdiff</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span><span class="w"> </span><span class="n">dups</span><span class="o">$</span><span class="n">inds2</span><span class="p">))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span></code></pre>
</div>
</div>
</li>
<li>
<p>Have a look at the <a href="https://link.springer.com/article/10.1007/BF00292322">graphical genotypes</a>.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot.geno.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">geno.image</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"white"</span><span class="p">,</span><span class="w"> </span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="s2">"violet"</span><span class="p">,</span><span class="w"> </span><span class="s2">"blue"</span><span class="p">))</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The default colors are red and green, which puts color blind people at a disadvantage. In this case, the genotypes AA, BB, AB are displayed in the colors red, blue, and violet, respectively. Missing data will be displayed in white.</td>
</tr>
</table>
</div>
</li>
<li>
<p>We remove the odd individuals with lots of missing calls.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">ntyped</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">sort</span><span class="p">()</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">rm.ind</span><span class="w">

</span><span class="n">subset</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">ind</span><span class="o">=</span><span class="n">setdiff</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">ntyped</span><span class="p">(</span><span class="n">mxb</span><span class="p">)),</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">rm.ind</span><span class="p">)))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">summary</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td><code>ntyped(mxb) |&gt; names()</code> outputs a list of all individuals.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Thin the set of markers.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">set.seed</span><span class="p">(</span><span class="m">1</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">lapply</span><span class="p">(</span><span class="n">pull.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="n">pickMarkerSubset</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">))</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">unlist</span><span class="p">()</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">keep</span><span class="w">

</span><span class="n">drop.markers</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">setdiff</span><span class="p">(</span><span class="n">markernames</span><span class="p">(</span><span class="n">mxb</span><span class="p">),</span><span class="w"> </span><span class="n">keep</span><span class="p">))</span><span class="w">  </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Markers are picked at random to have one per megabase. To ensure you always get the results, set the <a href="https://en.wikipedia.org/wiki/Random_seed">seed</a> for random number generator.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Look at the graphical genotypes again.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot.geno_1Mb.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">geno.image</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"white"</span><span class="p">,</span><span class="w"> </span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="s2">"violet"</span><span class="p">,</span><span class="w"> </span><span class="s2">"blue"</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
</li>
<li>
<p>Construct the map again.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">mstmap.cross</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">id</span><span class="o">=</span><span class="s2">"id"</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span><span class="w">
</span><span class="n">summary.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span></code></pre>
</div>
</div>
</li>
<li>
<p>Align the genetic to the physical map.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pdf</span><span class="p">(</span><span class="s2">"align_maps.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">alignCross</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">maps</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">mxb_physical</span><span class="p">),</span><span class="w"> </span><span class="n">layout</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
</li>
<li>
<p>Compute the rank correlation.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pull.map</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">as.table</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">a</span><span class="w">
</span><span class="n">pull.map</span><span class="p">(</span><span class="n">mxb_physical</span><span class="p">,</span><span class="w"> </span><span class="n">as.table</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">b</span><span class="w">
</span><span class="n">merge</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">m</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">sapply</span><span class="p">(</span><span class="n">split</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">m</span><span class="o">$</span><span class="n">chr.x</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="n">with</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">pos.x</span><span class="p">,</span><span class="w"> </span><span class="n">pos.y</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s1">'s'</span><span class="p">)))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">cc</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b>
</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td><code>by=0</code> means: merge by row names.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Use <code>method = 'p'</code> to compute the Pearson (linear) correlation.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Flip those linkage groups that are inverted relative to the reference orientation (short arm first).</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="nf">names</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">cc</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">flip.chr</span><span class="w">
</span><span class="n">flip.order</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">flip.chr</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span></code></pre>
</div>
</div>
</li>
<li>
<p>Plot the updated alignment and graphical genotypes.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">pdf</span><span class="p">(</span><span class="s2">"align_maps_2.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">alignCross</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">maps</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">mxb_physical</span><span class="p">),</span><span class="w"> </span><span class="n">layout</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot.geno.mstmap.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">geno.image</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"white"</span><span class="p">,</span><span class="w"> </span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="s2">"violet"</span><span class="p">,</span><span class="w"> </span><span class="s2">"blue"</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
</li>
<li>
<p>Compare genetic and physical map with connector plots.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">rescalemap</span><span class="p">(</span><span class="n">mxb_physical</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">/</span><span class="m">5</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb_physical</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_map_2.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">plotMap</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">mxb_physical</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Divide chromosome lengths by 5.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Save the genetic map object.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">copy</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb_genetic</span><span class="w">
</span><span class="n">saveRDS</span><span class="p">(</span><span class="n">mxb_genetic</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"mxb_genetic.Rds"</span><span class="p">)</span></code></pre>
</div>
</div>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_trait_mapping">7. Trait mapping</h2>
<div class="sectionbody">
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Calculate <a href="https://smcclatchy.github.io/mapping/03-calc-genoprob/">genotype probabilities</a> conditional on the marker data.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">calc.genoprob</span><span class="p">(</span><span class="n">mxb</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mxb</span></code></pre>
</div>
</div>
</li>
<li>
<p>Run the "QTL" scan for the first phenotypes, row type.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">mxb</span><span class="o">$</span><span class="n">pheno</span><span class="o">$</span><span class="n">row_type</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">mxb</span><span class="o">$</span><span class="n">pheno</span><span class="o">$</span><span class="n">row_type</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>

</span><span class="n">scanone</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">pheno.col</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"mr"</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="o">=</span><span class="s2">"binary"</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">out</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b>
</span><span class="n">summary</span><span class="p">(</span><span class="n">out</span><span class="p">)</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_row_type.pdf"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="3"></i><b>(3)</b>
</span><span class="n">plot</span><span class="p">(</span><span class="n">out</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Values for binary traits have to be zero or one.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Perform a single-QTL genome scan using the <a href="https://link.springer.com/article/10.1007/BF00223708">marker regression</a> (mr) method.
<a href="https://www.rdocumentation.org/packages/qtl/versions/1.47-9/topics/scanone">scanone()</a> supports several other methods and models. The defaut is "normal" for a quantiative phenotype.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>Plot the LOD (logarithm of the odds) scores along the genome.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Run a <a href="https://smcclatchy.github.io/mapping/06-perform-perm-test/">permutation test</a> to get p-values and a significance threshold.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">operm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scanone</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"mr"</span><span class="p">,</span><span class="w"> </span><span class="n">n.perm</span><span class="o">=</span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="n">pheno.col</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="n">operm</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">))</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_operm.pdf"</span><span class="p">)</span><span class="w"> <i class="conum" data-value="1"></i><b>(1)</b>
</span><span class="n">hist</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">operm</span><span class="p">))</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="n">summary</span><span class="p">(</span><span class="n">operm</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">)),</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s1">'red'</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="n">out</span><span class="p">,</span><span class="w"> </span><span class="n">perms</span><span class="o">=</span><span class="n">operm</span><span class="p">,</span><span class="w"> </span><span class="n">pvalues</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_row_type_threshold.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">out</span><span class="p">)</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">3.5</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s1">'red'</span><span class="p">)</span><span class="w"> <i class="conum" data-value="2"></i><b>(2)</b>
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>Plot a histogram of the maximum genome-wide LOD scores from the 1000 permutations.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>Add the significance threshold.</td>
</tr>
</table>
</div>
</li>
<li>
<p>Get interval estimates.</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">lodint</span><span class="p">(</span><span class="n">out</span><span class="p">,</span><span class="w"> </span><span class="n">chr</span><span class="o">=</span><span class="s2">"2H"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="o">=</span><span class="m">2</span><span class="p">)</span></code></pre>
</div>
</div>
</li>
<li>
<p>Plot the phenotypes against the genotypes at the most highly associated marker (genotype x phenotpe, pxg).</p>
<div class="listingblock">
<div class="content">
<pre class="rouge highlight"><code data-lang="r"><span class="n">rownames</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">out</span><span class="p">))</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">mar</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"plot_pxg.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">plotPXG</span><span class="p">(</span><span class="n">mxb</span><span class="p">,</span><span class="w"> </span><span class="n">pheno.col</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">marker</span><span class="o">=</span><span class="n">mar</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span></code></pre>
</div>
</div>
</li>
<li>
<p>To close R, type <code>quit()</code>. To close the tmux session, type <code>exit</code>. This will finish the tmux session. If you just want to detach and keep the session running, use <code>tmux detach</code>.</p>
</li>
<li>
<p>Check by BLAST how close the top marker is to the <a href="https://www.pnas.org/doi/10.1073/pnas.0608580104">causal gene</a> (<em>VRS1</em>). The sequence is available from <a href="https://www.ncbi.nlm.nih.gov/nuccore/MF776946.1">GenBank</a>.
GrainGenes offers <a href="https://wheat.pw.usda.gov">a web-based BLAST</a>.</p>
</li>
</ol>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2023-03-30 08:25:06 +0200
</div>
</div>
</body>
</html>